Taxonomy reorg per dewey decimal classifications (#1215)
instructlab/taxonomy: README.md

[1] Shivchander Sudalairaj*, Abhishek Bhandwaldar*, Aldo Pareja*, Kai Xu, David D. Cox, Akash Srivastava*. "LAB: Large-Scale Alignment for ChatBots", arXiv preprint arXiv: 2403.01081, 2024. (* denotes equal contributions)

## Choosing domains for the taxonomy

> In general, we use the Dewey Decimal Classification (DDC) System to determine our domains (and subdomains) in the taxonomy. This SUMMARIES of the DDC document is a great resourcce for determining where a topic might be classified:

Reviewer: ```suggestion
In general, we use the Dewey Decimal Classification (DDC) System to determine our domains (and subdomains) in the taxonomy. The [DDC SUMMARIES document](https://www.oclc.org/content/dam/oclc/dewey/resources/summaries/deweysummaries.pdf) is a great resource for determining where a topic might be classified.
```

> https://www.oclc.org/content/dam/oclc/dewey/resources/summaries/deweysummaries.pdf

Reviewer: ```suggestion
```

> If you are unsure where to put your knowledge or compositional skill, create a folder in the `miscellaneous_unknown` folder.

Reviewer: ```suggestion
If you are unsure where to put your knowledge or compositional skill, create a folder in the `miscellaneous_unknown` folder somewhere under the `knowledge` or `compositional_skills` folders.
```

> There is a limit to how much content can exist in the question/answer pairs for the model to process. Due to this, only add a maximum
> of around 2300 words to your question and answer seed example pairs in the `qna.yaml` file.

> Compositional skills can either be grounded (includes a context) or ungrounded (does not include a context).  Grounded or ungrounded is the last node in the taxonomy tree, for example: `linguistics/writing/poetry/haiku/ungrounded` or `linguistics/grammar/grounded`. The `qna.yaml` is in this final node.

Reviewer: We need to strike the reference to ungrounded in the path here.

Reviewer 2: updated.

> There is a limit to how much content can exist in the question/answer pairs for the model to process. Due to this, only add a maximum
> of around 2300 words to your question and answer seed example pairs in the `qna.yaml` file.

> Compositional skills can either be grounded (includes a context) or ungrounded (does not include a context).  Grounded or ungrounded is the last node in the taxonomy tree, for example: `linguistics/writing/poetry/haiku/ungrounded` or `linguistics/grammar/grounded`. The `qna.yaml` is in this final node.

Reviewer: Again, I am unclear on the utility of putting grounded/ungrounded in the path (since the qna.yaml can have or not have context independent of the path).

- `seed_examples`: A collection of key/value entries. New
  submissions should have at least five entries, although
  older files may have fewer. **Required**

>   - `context`: Grounded skills require the user to provide context containing information that the model is expected to take into account during processing. This is different from knowledge, where the model is expected to gain facts and background knowledge from the tuning process. The context key is optional for ungrounded skills.

Reviewer: ```suggestion
  - `context`: Grounded skills require the user to provide context containing information that the model is expected to take into account during processing. This is different from knowledge, where the model is expected to gain facts and background knowledge from the tuning process. The context key should not be used for ungrounded skills.
```

itself, plus any added directories that contain the file, is the entirety of the skill
in terms of a taxonomy contribution:

#### Ungrounded compositional skill: Directory tree example

```ascii
[...]

└── writing
    └── poetry
    |   └── haiku <=== here it is :)

>     |   |   └── ungrounded

Reviewer: No ungrounded

    |   |   |    └── qna.yaml
    |   |            attribution.txt
        [...]
    └── prose
    |   └── debate

>     |   |   └── ungrounded

Reviewer: No ungrounded

> [!IMPORTANT]
> Upon release, the taxonomy repository is only accepting contributions from Wikipedia, Project Gutenberg, Wikisource (library), OpenStax textbooks family of publications, The Open Organization publications, and The Scrum Guide. and is capped at 50 contributions. If you want to add knowledge to the taxonomy repository, please fill out this [InstructLab Knowledge Submission Registration](https://docs.google.com/forms/d/1VWJ_XPwH3gBTIXCabpWc0I5pjWIlXETMSFKXc8fpgkA/viewform?edit_requested=true) form and await acceptance! Please do not add contributions if you do not receive the confirmation email. Thank you!

While skills are foundational or performative, knowledge is based more on answering questions that involve facts, data, or references.

> Knowledge is supported by documents, such as a textbook, technical manual, encyclopedia, journal, or magazine. The document type is the last node in the taxonomy tree, for example: `science/animals/tabby_cat/encyclopedia`. The `qna.yaml` is in this final node.

Reviewer: ```suggestion
Knowledge is supported by documents, such as a textbook, technical manual, encyclopedia, journal, or magazine.
```

> [!IMPORTANT]
> Upon release, the taxonomy repository is only accepting contributions from Wikipedia and is capped at 50 contributions. If you want to add knowledge to the taxonomy repository, please fill out this [InstructLab Knowledge Submission Registration](https://docs.google.com/forms/d/1VWJ_XPwH3gBTIXCabpWc0I5pjWIlXETMSFKXc8fpgkA/viewform?edit_requested=true) form and await acceptance! Please do not add contributions if you do not receive the confirmation email. Thank you!

While skills are foundational or performative, knowledge is based more on answering questions that involve facts, data, or references.

> Knowledge is supported by documents, such as textbook, technical_manual, encyclopedia, journal, or magazine. The document type is the last node in the taxonomy tree, for example: `science/animals/tabby_cat/encyclopedia`. The `qna.yaml` is in this final node.

Reviewer: ```suggestion
Knowledge is supported by documents, such as a textbook, technical manual, encyclopedia, journal, or magazine. The document type is the last node in the taxonomy tree, for example: `science/animals/tabby_cat/encyclopedia`. The `qna.yaml` is in this final node.
```

Also, not sure of the utility of putting the doc type in the path. It is also possible the knowledge documents come from multiple sources: web page, pdf document, wikipedia article.

Author: @bjhargrave This was previously discussed and agreed upon that it would be the final node in the path.  I would very much like to see this JUST in the qna.yaml file in the document: section, but the majority of folks said that they'd like to see it in the taxonomy tree for when they just scan/read the tree.

Reviewer: I would like to open this up for discussion again :-)

domain. Maintainers can decide to change the names of the existing branches or to add new branches.

> [!IMPORTANT]
> Folder names do not have spaces. Use underscores between words.

Below is an illustrative directory structure to show this layout:

```ascii
.
└── linguistics
    ├── writing
    │   ├── brainstorming
    │   │   ├── idea_generation

>     |   |   |   └── ungrounded

Reviewer: no ungrounded

    │   │   │       └── qna.yaml
    │   │   │           attribution.txt
    │   │   ├── refute_claim

>     |   |   |   └── ungrounded

Reviewer: no ungrounded

    │   │   │       └── qna.yaml
    │   │   │           attribution.txt
    │   ├── prose
    │   │   ├── articles

>     |   |   |   └── ungrounded

Reviewer: no ungrounded
